Related to issue #5876: reduce excessive copying of message data

plubius opened the pull request:
Many .png tile messages are very large so avoid using -[NSString stringByAppendingString:] calls to create the JavaScript string.

mmeeks reviewed (approved):
Looks great =) I wonder if we could pre-allocate std::vector<char> data - to avoid lots of reallocs as we push lots of stuff to it; since it's not long-lived a larger alloc might be good - length * 4 or whatever ? =)

plubius commented:
> Looks great =) I wonder if we could pre-allocate std::vector data - to avoid lots of reallocs as we push lots of stuff to it; since it's not long-lived a larger alloc might be good - length * 4 or whatever ? =)

Good idea. I've added a new commit to this pull request. I did some printfs and it appears that binary messages consistently use up at least 75% of (length * 4) + 1.

mmeeks commented on common/MobileApp.cpp:
```diff
@@ -38,7 +38,10 @@ DocumentData & DocumentData::get(unsigned docId)
     assert(idToDocDataMap.find(docId) != idToDocDataMap.end());
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused"
     DocumentData &d = get(docId);
```
Urk, all these pragmas make for grim reading & maintainability. Can't we just delete this redundant line ? =) Ditto for other places where things are not needed (?) What are all the:

#pragma clang diagnostic ignored "-Wconditional-uninitialized"

warnings in fact ? - can we do something more elegant there; we really want fewer conditionals & pre-processing in the code if we can.

plubius replied:
OK. I was able to remove all the #pragma clang's as I found the build setting in Xcode that was causing this: I changed "Uninitialized Variables" from "Yes (Aggressive)" to "Yes".

I also added comments to the two cases that have odd code that generated warnings: the destructor case and the "comma in an if statement" case. Maybe they are cruft but maybe these specific cases were intentional so better to be safe and leave the code as is but without warnings.

Overall my goal was to cut down the number of warnings to a very small number. I found that several deprecation warnings were being missed because they scrolled out of view in Xcode and since Apple routinely deprecates things, I wanted to get those pushed to the top of the build warnings list.

mmeeks reviewed (commented):
Reserving looks awesome =) thanks !

plubius commented:
OK. I amended the last patch to revert all of the #pragma clang's. I found the "Uninitialized Variable" build setting in Xcode and changed it from "Yes (Aggressive)" to "Yes" and that removes the false positive uninitialized warnings.

The other two warning cases I solved by slight modifications to the code but without any logic changes and added a comment to each.

I didn't want to clean up the last two warning cases because, although they appear odd, my goal was only to eliminate the very large number of Xcode warnings. Because of all the noise, I found that I was not seeing several deprecated warnings. Apple deprecates things constantly and many deprecations turn into crashes on newer OS versions so I wanted to increase the visibility of deprecated warnings without changing any code.

mmeeks reviewed (approved):
Looks great - thanks Patrick ! =)
